### DISCUSSION TITLE
[Question] Recommend Configuration & Usage EntityManager with Nest.js

### DISCUSSION THREAD

**Discussion by jays1204** [mikro-orm/nestjs#69] [Q&A] [answered]
I want to know how to use it properly.
With Nest.js, I configure RequestContext on middleware. Because this [document](https://mikro-orm.io/docs/identity-map#-requestcontext-helper-for-di-containers.) is recommend to be set like below.

part of my configuration.
- main.ts
```
app.use((req, res, next) => {
  RequestContext.create(orm.em, next);
});
```
- app.module.ts
```
@Module({
  imports: [
    MikroOrmModule.forRoot({
      // ...
      registerRequestContext: false,
      //....
    }),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
```

Here's the question.

I want to use EntityManager On Service Class(Provider).
I would like to know which of the two below is the more recommended way.

1. EntityManager by DI
```
import { MikroORM } from '@mikro-orm/core';
import { EntityManager } from '@mikro-orm/mysql'; // Import EntityManager from your driver package or `@mikro-orm/knex`

@Injectable()
export class MyService {

  constructor(private readonly orm: MikroORM,
              private readonly em: EntityManager) {
  }

  async someMethod(): {
        const post = await this.em.find(Post, ...);
  }
}
```

2.  EntityManager by RequestContext
```
@Injectable()
export class MyService {

  constructor() {  }

  async someMethod(): {
    const em  = RequestContext .getEntityManager();
    const post = await em.find(Post, ...);
  }
}

```

I think 2 is proper way than 1. Because `RequestContext .getEntityManager()` fork by each http request.
1 is re-use entityManger not fork. https://github.com/mikro-orm/nestjs/blob/7dfc0975523c1abe33bd6302237f1719e12fe4d5/src/mikro-orm.providers.ts#L42

Is it right?

**Comment by B4nan** [comment, marked as answer]
Both ways are doing the same, and actually the first one should be preferred. The point of using `RequestContext` helper is to allow the 1. way. We still work with the fork internally, as all methods on the EM that are using the context somehow will first get the fork via `em.getContext()`, which is by default doing `RequestContext.getEntityManager()`. You can check the `em.id`, it should show you different values for each request, even if you use the instance from DI. Those instances in DI are always global, but when they are executed inside context handler (so inside the middleware handler), they will see the fork and use that instead. With v5 we even validate that you are not touching the global context anyhow. In other words, you will get an error if the middleware is not working for some reason.

**Reply by jays1204** [reply]
Thank you!
